command/apply.go

```diff
@@ -22,105 +21,177 @@ type ApplyCommand struct {
+	diags = diags.Append(beDiags)
+	if diags.HasErrors() {
+		view.Diagnostics(diags)
+		return 1
 	}
 
-	configPath, err := ModulePath(args)
+	// Build the operation request
+	opReq, opDiags := c.OperationRequest(be, view, planFile, args.Operation)
+	diags = diags.Append(opDiags)
+
+	// Collect variable value and add them to the operation request
+	diags = diags.Append(c.GatherVariables(opReq, args.Vars))
+
+	// Before we delegate to the backend, we'll print any warning diagnostics
+	// we've accumulated here, since the backend will start fresh with its own
+	// diagnostics.
+	view.Diagnostics(diags)
+	if diags.HasErrors() {
+		return 1
+	}
+	diags = nil
+
+	// Run the operation
+	op, err := c.RunOperation(be, opReq)
 	if err != nil {
-		c.Ui.Error(err.Error())
+		diags = diags.Append(err)
+		view.Diagnostics(diags)
 		return 1
 	}
 
-	// Check for user-supplied plugin path
-	if c.pluginPath, err = c.loadPluginPath(); err != nil {
-		c.Ui.Error(fmt.Sprintf("Error loading plugin path: %s", err))
+	if op.Result != backend.OperationSuccess {
+		return op.Result.ExitStatus()
+	}
+
+	// // Render the resource count and outputs
+	view.ResourceCount()
+	if !c.Destroy && op.State != nil {
+		view.Outputs(op.State.RootModule().OutputValues)
+	}
+
+	view.Diagnostics(diags)
+
+	if diags.HasErrors() {
 		return 1
 	}
 
-	// Try to load plan if path is specified
+	return 0
+}
+
+func (c *ApplyCommand) LoadPlanFile(path string) (*planfile.Reader, tfdiags.Diagnostics) {
 	var planFile *planfile.Reader
-	if planPath != "" {
-		planFile, err = c.PlanFile(planPath)
+	var diags tfdiags.Diagnostics
+
+	// Try to load plan if path is specified
+	if path != "" {
+		var err error
+		planFile, err = c.PlanFile(path)
 		if err != nil {
-			c.Ui.Error(err.Error())
-			return 1
+			diags = diags.Append(tfdiags.Sourceless(
+				tfdiags.Error,
+				fmt.Sprintf("Failed to load %q as a plan file", path),
+				fmt.Sprintf("Error: %s", err),
+			))
+			return nil, diags
 		}
 
 		// If the path doesn't look like a plan, both planFile and err will be
 		// nil. In that case, the user is probably trying to use the positional
 		// argument to specify a configuration path. Point them at -chdir.
 		if planFile == nil {
-			c.Ui.Error(fmt.Sprintf("Failed to load %q as a plan file. Did you mean to use -chdir?", planPath))
-			return 1
+			diags = diags.Append(tfdiags.Sourceless(
+				tfdiags.Error,
+				fmt.Sprintf("Failed to load %q as a plan file", path),
+				"The specified path is a directory, not a plan file. You can use the global -chdir flag to use this directory as the configuration root.",
+			))
+			return nil, diags
 		}
 
 		// If we successfully loaded a plan but this is a destroy operation,
 		// explain that this is not supported.
 		if c.Destroy {
-			c.Ui.Error("Destroy can't be called with a plan file.")
-			return 1
-		}
-	}
-	if planFile != nil {
-		// Reset the config path for backend loading
-		configPath = ""
```

Comment:
Good question! This reset was already irrelevant in the old code. `configPath` used to really be `maybePlanFileOrMaybeConfigPath`, so we needed to reset it if a plan file was loaded. Now the reset does nothing, as `configPath` is now always the post-`chdir` working-directory.

Just to dig into this a bit more, here's a trace of where `configPath` goes on the main branch:

https://github.com/hashicorp/terraform/blob/7b7273e3aafb1f458e72735aae491dfa24786b91/command/apply.go#L119
https://github.com/hashicorp/terraform/blob/7b7273e3aafb1f458e72735aae491dfa24786b91/command/meta_config.go#L150-L151
https://github.com/hashicorp/terraform/blob/7b7273e3aafb1f458e72735aae491dfa24786b91/command/meta_config.go#L79-L81
https://github.com/hashicorp/terraform/blob/7b7273e3aafb1f458e72735aae491dfa24786b91/command/meta_config.go#L29-L51

And because it's not trivially obvious to me what `normalizePath` does, [here's a playground showing that `normalizePath("")` results in `"."`](https://play.golang.org/p/E7K0mBUF_at).